plugins/forwarder/kafka/nativetracing/forwarder.go

```diff
@@ -0,0 +1,121 @@
+// Licensed to Apache Software Foundation (ASF) under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Apache Software Foundation (ASF) licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package nativetracing
+
+import (
+	"fmt"
+	"reflect"
+	v3 "skywalking.apache.org/repo/goapi/collect/language/agent/v3"
+
+	"github.com/Shopify/sarama"
+	"github.com/apache/skywalking-satellite/internal/pkg/config"
+	"github.com/apache/skywalking-satellite/internal/satellite/event"
+	"google.golang.org/protobuf/proto"
+
+	v1 "skywalking.apache.org/repo/goapi/satellite/data/v1"
+)
+
+const (
+	Name     = "native-tracing-kafka-forwarder"
+	ShowName = "Native Tracing Kafka Forwarder"
+)
+
+type Forwarder struct {
+	config.CommonFields
+	Topic    string `mapstructure:"topic"` // The forwarder topic.
+	producer sarama.SyncProducer
+}
+
+func (f *Forwarder) Name() string {
+	return Name
+}
+
+func (f *Forwarder) ShowName() string {
+	return ShowName
+}
+
+func (f *Forwarder) Description() string {
+	return "This is a synchronization Kafka forwarder with the SkyWalking native log protocol."
+}
+
+func (f *Forwarder) DefaultConfig() string {
+	return `
+# The remote topic. 
+topic: "skywalking-segments"
+`
+}
+
+func (f *Forwarder) Prepare(connection interface{}) error {
+	client, ok := connection.(sarama.Client)
+	if !ok {
+		return fmt.Errorf("the %s only accepts a grpc client, but received a %s",
+			f.Name(), reflect.TypeOf(connection).String())
+	}
+	producer, err := sarama.NewSyncProducerFromClient(client)
+	if err != nil {
+		return err
+	}
+	f.producer = producer
+	return nil
+}
+
+func (f *Forwarder) Forward(batch event.BatchEvents) error {
+	var message []*sarama.ProducerMessage
+	for _, e := range batch {
+		switch data := e.GetData().(type) {
+		case *v1.SniffData_Segment:
+			segmentObject := &v3.SegmentObject{}
+			err := proto.Unmarshal(data.Segment, segmentObject)
+			if err != nil {
+				return err
+			}
+
+			message = append(message, &sarama.ProducerMessage{
+				Topic: f.Topic,
+				Key:   sarama.StringEncoder(segmentObject.GetTraceSegmentId()),
+				Value: sarama.ByteEncoder(data.Segment),
+			})
+		case *v1.SniffData_SpanAttachedEvent:
+			message = append(message, &sarama.ProducerMessage{
```

Comment:
Today I was looking for the entity of SniffData_SpanAttachedEvent corresponding to the code of a Java agent, but I couldn't find it. I also tried to debug the project with breakpoints, but it didn't go into this part of the code. Can we ignore this part of the code for now, or could you give me some hints on how to locate it? Thank you.   @mrproliu

Reply:
`SniffData_SpanAttachedEvent` is from ebpf agent, skywalking-rover project.

Reply:
You could find it here, https://github.com/apache/skywalking-data-collect-protocol/blob/0da9c8b3e111fb51c9f8854cae16d4519462ecfe/language-agent/Tracing.proto#L244

Reply:
I carefully looked into it and we might not be able to add the SpanAttachedEvent feature to OAP anytime soon. So, for now, we can simply warn users that the current version of Kafka  forwarder  does not support this feature. It's not the best solution, but it's the only option we have for now.

Reply:
What blocks you? I think this is very easy to add. It is just a part of the trace. Routing like a segment should be enough.

Reply:
I am not familiar with Rover and  cannot conduct testing. If I were to proceed without proper knowledge, it could negatively impact the quality.

Reply:
You don't need to understand all. You just need to copy and change a case of it.
Take this, https://github.com/apache/skywalking/blob/2e04b069bfba24b1233dd66588217d08bdde7710/.github/workflows/skywalking.yaml#L523-L524.

Verify relative traces, but may be not real attached events.

https://github.com/apache/skywalking/blob/master/test/e2e-v2/cases/profiling/ebpf/network/network-cases.yaml#L143-L156

@mrproliu What do you want to add?

Reply:
This is not resolved.

Reply:
I'm not quite sure if it's like this. If it is, I'll continue tomorrow when I have time. Thank you. 
https://github.com/zbw911/skywalking-satellite/blob/e86c410f1d06961f9b4be0dd26131772245f1ed2/plugins/forwarder/kafka/nativetracing/forwarder.go#L94

Reply:
> Why does the attached event hasn't a key of the message? Also, will you create a PR for handling the attached event in the OAP?

A TracingType currently covers two types of business requests, which has led to the current if-else problem. The setting of SniffType needs to be further detailed and each type should be a type of business.

Types should be defined as more specific types rather than classifications. SniffType should be mapped to the service layer in proto.

Finally, my solution is to add SpanAttachedEventType = 17 in https://github.com/apache/skywalking-goapi/blob/main/satellite/data/v1/SniffData.proto#L47.

If possible, I will branch the goapi project, and then modify the skywalking-satellite project.

@mrproliu  

-------
v1.SniffData_SpanAttachedEvent->SpanAttachedEventReportService：  https://github.com/apache/skywalking-data-collect-protocol/blob/0da9c8b3e111fb51c9f8854cae16d4519462ecfe/language-agent/Tracing.proto#L233-L237

v1.SniffData_Segment->TraceSegmentReportService：  https://github.com/apache/skywalking-data-collect-protocol/blob/0da9c8b3e111fb51c9f8854cae16d4519462ecfe/language-agent/Tracing.proto#L33-L46

Reply:
Why do you think the `if-else` is a problem? They all belong to the tracing data. For this review, the question is the Attached span hasn't a message key, not another question.

Reply:
The existing OAP server cannot parse SpanAttachedEvent (https://github.com/apache/skywalking/blob/2e04b069bfba24b1233dd66588217d08bdde7710/oap-server/server-fetcher-plugin/kafka-fetcher-plugin/src/main/java/org/apache/skywalking/oap/server/analyzer/agent/kafka/provider/handler/TraceSegmentHandler.java#L71-L86).

Parsing in the current mode may look like the following:
``` 
try{
        SegmentObject segment = SegmentObject.parseFrom(record.value().get());
        segmentParserService.send(segment);
}
catch(InvalidProtocolBufferException e)
{
      try{
          SpanAttachedEvent spanAttachedEvent = SpanAttachedEvent.parseFrom(record.value().get());
          //.................
      }
      catch(InvalidProtocolBufferException e )
      {
      // ...........
      }
}
```

Making changes to the above code is somewhat difficult.

Therefore, a warning message should be added to the tracing forwarder, explicitly indicating that the current Kafka mode does not support SpanAttachedEvent, so that users can be aware of this limitation and take appropriate action.

Thus, instead of adding a key to SpanAttachedEvent, we should explicitly inform the user that the current Kafka mode does not support this event to prevent misuse.

If necessary, we can increase the log level to error.

Additionally, I suggest refining the type in goapi as a possible future option.

Reply:
I think the best approach for now is to temporarily abandon support for *v1.SniffData_SpanAttachedEvent in Kafka mode.

Reply:
Are you really reading the protocol? Read more carefully, rather than guessing.

https://github.com/apache/skywalking-data-collect-protocol/blob/0da9c8b3e111fb51c9f8854cae16d4519462ecfe/language-agent/Tracing.proto#L233-L237

What you thought is actually impossible from gRPC stand point.

Reply:
All protocols are only getting accepted if the backend supports it. 
Be clear about the status, and be patient in reading projects and docs. 

You are proposing a big change. From a maintainer perspective, it is already not easy to follow your first contribution as huge, you should get familiar with the project more rather than keeping saying `best approach` or `should`.

Again and again, we have given you the e2e to simulate how to use ebpf agent and all these protocols. You at least should follow them and try them.

What you replied makes me concerned you are only pushing this in hurry, rather than really want to understand what is the case.

Reply:
@mrproliu Please be very careful about all the codes this PR is supposed to change. I am super concerned about it lacking fully testing

Reply:
@zbw911 please finish this issue first, then click the `resolve conversation` button. 

> The existing OAP server cannot parse SpanAttachedEvent

Back to my first comment on this conversation, you should also implement the relevant functions on the OAP side.
You have already implemented the management Kafka forwarder, so you should know there are also multiple data structures in that service. In that implementation, you already use the `different message key prefixes` to identify them, you can do this in the same way. 
You should read the OAP code related to the management Kafka receiver: https://github.com/apache/skywalking/blob/master/oap-server/server-fetcher-plugin/kafka-fetcher-plugin/src/main/java/org/apache/skywalking/oap/server/analyzer/agent/kafka/provider/handler/ServiceManagementHandler.java#L61-L71